pkg/backends/calico/client.go

```diff
@@ -207,50 +481,95 @@ func (c *client) OnUpdates(updates []api.Update) {
+		if !ok {
+			// Not a v3 resource.
 			continue
 		}
 
-		switch u.UpdateType {
-		case api.UpdateTypeKVDeleted:
-			delete(c.cache, k)
-		case api.UpdateTypeKVNew, api.UpdateTypeKVUpdated:
-			value, err := model.SerializeValue(&u.KVPair)
+		if v3key.Kind == apiv3.KindNode {
+			// Convert to v1 key/value pairs.
+			log.Debugf("Node: %#v", u.Value)
+			if u.Value != nil {
+				log.Debugf("BGPSpec: %#v", u.Value.(*apiv3.Node).Spec.BGP)
+			}
+			kvps, err := c.nodeV1Processor.Process(&u.KVPair)
 			if err != nil {
-				log.Errorf("Ignoring update: unable to serialize value %v: %v", u.KVPair.Value, err)
+				log.Errorf("Problem converting Node resource: %v", err)
 				continue
 			}
-			c.cache[k] = string(value)
-		}
+			for _, kvp := range kvps {
+				log.Debugf("KVP: %#v", kvp)
+				if kvp.Value == nil {
+					c.updateCache(api.UpdateTypeKVDeleted, kvp)
+				} else {
+					c.updateCache(u.UpdateType, kvp)
+				}
+			}
 
-		log.Debugf("Cache entry updated from event type %d: %s=%s", u.UpdateType, k, c.cache[k])
-		if c.synced {
-			c.keyUpdated(k)
+			// Update our cache of node labels.
+			v3res, ok := u.Value.(*apiv3.Node)
+			if !ok {
+				log.Warning("Bad value for Node resource")
+				continue
+			}
+			if v3res != nil {
+				c.nodeLabels[v3key.Name] = v3res.Labels
+			} else {
+				delete(c.nodeLabels, v3key.Name)
+			}
+
+			// Note need to recompute BGP v1 peerings.
+			needUpdatePeersV1 = true
 		}
-	}
 
-	if c.synced {
-		// If we're not running for a route reflector, and the node-to-node mesh
-		// configuration has been toggled, we need to terminate so that confd can be
-		// restarted and monitor the correct set of nodes.  (If we running for a route
-		// reflector, we always monitor all nodes.)
-		if !c.routeReflector {
-			// The node is disabled if the setting is present and is set to false.  Although this
-			// is a json blob, it only contains a single field, so searching for "false" will
-			// suffice.
-			nodeMeshEnabled := !strings.Contains(c.cache[globalNodeMesh], "false")
-			if nodeMeshEnabled != c.nodeMeshEnabled {
-				log.Info("Node to node mesh setting has been modified - restarting confd")
-				os.Exit(1)
+		if v3key.Kind == apiv3.KindBGPPeer {
+			// Update our cache of v3 BGPPeer resources.
+			if u.Value == nil || u.UpdateType == api.UpdateTypeKVDeleted {
+				delete(c.bgpPeers, v3key.Name)
+			} else if v3res, ok := u.Value.(*apiv3.BGPPeer); ok {
+				c.bgpPeers[v3key.Name] = v3res
+			} else {
+				log.Warning("Bad value for BGPPeer resource")
+				continue
 			}
+
+			// Note need to recompute equivalent v1 peerings.
+			needUpdatePeersV1 = true
 		}
+	}
+
+	// If configuration relevant to BGP peerings has changed, recalculate the set of v1
+	// peerings that should exist, and update the cache accordingly.
+	if needUpdatePeersV1 {
+		log.Debug("Recompute BGP peerings")
+		c.updatePeersV1()
+	}
 
+	// Update our cache from each of the individual updates, and keep track of
```

Comment:
Yes, correct.  With the v1 paths as before, and then with the v3 path `/calico/resources/v3/projectcalico.org/nodes/<name>`.

In theory a custom template could make use of the v3 node, but perhaps that's too much of a stretch in practice and so it would be better to suppress adding those to the cache.  WDYT?

Reply:
Personally, I'd prefer deprecating the v1 paths and moving to v3 entirely, but obviously that requires more thinking before we can just do it.

I think I'm OK including both, though it will have some minimal occupancy impact.

Reply:
OK cool, thanks.